utils/scripts/helpers/install_maldet.sh

```diff
@@ -0,0 +1,59 @@
+#!/bin/bash
+
+# Installs maldet, clamAV and uses the current Jenkins job
+# as an artifact server.
+#
+# Makes an effort to update the current maldet database.
+# it is not fatal if the maldet database can not be updated.
+
+set -uex
+
+mydir="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+# shellcheck source=utils/scripts/helpers/distro_info.sh
+source "$mydir/distro_info.sh"
+
+# We need sudo for running the scan and git for backward
+# compatibility.
+if command -v dnf; then
+  dnf install clamav clamav-devel git gzip hostname sudo
+  # Some Dockerfiles missing the md5sum command by not having
+  # coreutils installed.   Rocky 8 pre-installs coreutils-single which
+  # blocks coreutils from being installed.
+  if ! command -v md5sum; then
+      dnf install coreutils
+  fi
+elif command -v apt-get; then
+  apt-get --assume-yes install clamav libclamav-dev \
+                       coreutils git gzip hostname sudo
+  service clamav-freshclam stop || true
+  systemctl disable clamav-freshclam || true
+fi
+mkdir -p /etc/clamd.d
+printf "LogSyslog yes\n" >> /etc/clamd.d/scan.conf
+
+lmd_tarball='maldetect-current.tar.gz'
+: "${JENKINS_URL:=https://build.hpdd.intel.com/}"
+
+lmd_url="${JENKINS_URL}job/daos-stack/job/tools/job/master/"
+lmd_url+="lastSuccessfulBuild/artifact/${lmd_tarball}"
+curl "${lmd_url}" --silent --show-error --fail -o "/var/tmp/${lmd_tarball}"
+
+lmd_src='lmd_src'
+mkdir -p "/var/tmp/${lmd_src}"
+tar -C "/var/tmp/${lmd_src}" --strip-components=1 -xf "/var/tmp/${lmd_tarball}"
+pushd "/var/tmp/${lmd_src}"
+  ./install.sh
+popd
+
+/usr/local/sbin/maldet --update-sigs || true
```

Comment:
I still think this is wrong as it'll mean two things:

1. The checks that get run will be inconsistent from run to run based on if this failed or not, and we will have no way of knowing if a job used stale checks or up-to-date ones.
2. If there's a more permanent error here (maldet is misconfigured or networking issues etc) this could become permanently disabled and we'd have no indication of this, leading us to think we were getting up-to-date checking when in fact we were not.

Reply:
I have to agree that ignoring a failure here of `maldet` really concerns me as well.

Reply:
As I have stated before we have these choices:

1. Not run any malware scans until we actually ship the product and cause everything to be held up then.
2. Have CI fail for days the next time that the maldet signature server goes offline as it did which triggered this stage being disabled by default.
3. Accept that we have no control over that server or any way to proxy or cache the maldet data.
4. Reverse engineer how maldet downloads the signature updates and locally patch maldet to use the cached data..

Again the maldet signature update is an update to the a set of signatures in the maldet tarball and the signatures from the ClamAV product, and we are locally caching the ClamAV signatures, so the malware scans are not totally disabled if we are not getting maldet signature updates.

This is intended to give us a best effort early warning of any potential issues.   We run a separate maldet scan at the time we sign packages for releases.

Reply:
Can we run maldet on daily or master builds only and have it configured to continue-on-failure of this stage?  That way we'd be notified of any regression or upstream failure but if it happened often we would not block development progress.

Reply:
Or can we proxy the upstream data in an artifactory _remote_, _Generic_ proxy?

But perhaps @ashleypittman's suggestion is actually better.  We need visibility of the stage failing, but we don't want to stop PRs (or landings and dailies for that matter) on that failing stage.  I.e. make the stage unstable but allow continuing so that the pipeline finishes but is unstable.

Reply:
More people are looking at PR runs.   The pipeline-lib change needed to enable it again can be set on a branch specific basis.

I do not think I can make a failure in a docker build be caught to convert it to be a continue on failure of this stage.

Reply:
As I understand it, the way it currently sits is this:

- We will be able to enable the maldet scanning again.

- We will be able to continue (no downtime) if a maldet upstream failure occurs.

- If an upstream maldet failure occurs, we will be able to continue scanning for known maldet issues, but we won't receive new signatures.  
    - In that case, we could potentially pass PRs through that /should/ have been flagged ( again, if there was a new signature that we missed and that PR fit that new signature)
      - If those PRs do slip through, that issue will be caught in master/daily scans as soon as the maldet upstream issue is fixed and we pick up the new signatures.
      
That sounds like an acceptable risk to me.  Am I missing something?

Reply:
> If those PRs do slip through, that issue will be caught in master/daily scans as soon as the maldet upstream issue is fixed and we pick up the new signatures.

But how do we ever know the upstream is broken and then fixed again if the failure is always ignored and always passes the stage regardless of whether the `maldet` passed or failed on either PRs or master/daily.  Upstream could get broken indefinitely (i.e. some event that requires action on our side for example) and we'd never know.  Note that PRs and master/daily are going to fail silently also.  I don't see any special handling of the ignored `maldet` failure for master/daily runs.

Ultimately, we will never know for a given run's (including release candidate builds!) artifacts, if they have a virus in them or not without manually inspecting the scanning stage for that run to see if the `true` was executed.  Nobody is ever going to do that.

Reply:
What @brianjmurrell said, with the extra that because of docker caching we'll have no way of knowing if the `true` was executed as the image will come from the cache and we'll no longer have the logs for it.

The concern here is that if it fails for whatever reason then it'll fail silently and we'll never know.

Reply:
> What @brianjmurrell said, with the extra that because of docker caching we'll have no way of knowing if the `true` was executed as the image will come from the cache and we'll no longer have the logs for it.

The `maldet` is done in the `docker run` phase not the `docker build` phase so the layer caching won't hide it at least.  But it is still buried in a log that nobody will look at because the stage is green.
 
> The concern here is that if it fails for whatever reason then it'll fail silently and we'll never know.

Indeed, that is still the concern.  It will never show up.  Anywhere.  Not in PRs, not master, not daily.  Nowhere.